### PR TITLE
Add table function generate_subscripts

### PR DESCRIPTION
**Pull request by marregui**
Convenience function that generates the set of valid subscripts for the
specified dimension of the given array. Zero rows are returned for arrays
that do not have the requested dimension, or for NULL arrays (but valid
subscripts are returned for NULL array elements).

### CODE REVIEW

File: docs/general/builtins/table-functions.rst [hunk 1]
```diff
@@ -140,6 +140,54 @@ The return value always matches the ``start`` / ``stop`` types.
...
+Generate the subscripts for the specified dimension ``dim`` of the given
+``array``. Zero rows are returned for arrays that do not have the requested
+dimension, or for NULL arrays (but valid subscripts are returned for NULL
+array elements).
+
+If ``reverse`` is ``true`` the subscripts will be returned in reverse order.
+
+::
+
+    cr> SELECT generate_subscripts([NULL,1,NULL,2], 1) AS s;
```
Comment (mechanomi) [inline docs/general/builtins/table-functions.rst:157]:
```suggestion
    cr> SELECT generate_subscripts([NULL, 1, NULL, 2], 1) AS s;
```

File: docs/general/builtins/table-functions.rst [hunk 2]
```diff
@@ -140,6 +140,54 @@ The return value always matches the ``start`` / ``stop`` types.
...
+    | 1 |
+    | 2 |
+    | 3 |
+    | 4 |
+    +---+
+    SELECT 4 rows in set (... sec)
+
+::
+
+    cr> SELECT generate_subscripts([NULL,1,NULL,2], 1, true) AS s;
```
Comment (mechanomi) [inline docs/general/builtins/table-functions.rst:170]:
```suggestion
    cr> SELECT generate_subscripts([NULL, 1, NULL, 2], 1, true) AS s;
```

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        List<?> nestedArray = extractNestedArray(array, dim);
+        if (nestedArray == null || nestedArray.isEmpty()) {
+            return () -> NO_RESULTS_ITERATOR;
+        }
+        boolean reversed = args.length == 3 && (Boolean) args[2].value();
+        return () -> new Iterator<>() {
+
+            final Object [] columns = new Object[]{ null };
+            final RowN row = new RowN(columns);
+            final Iterator<Integer> subscripts = IntStream
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:127]: I am not sure that stream usage is justified here. It generates the boxed sequence, sorts it if needed, allocates additional unnecessary objects, etc. As an idea, can't we just have two int fields, e.g. `from` and `to` that we set depending on the reverse flag, also probably we need the sign, so we know whether we inc or dec or smth similar? Something like that will be definitely more efficient than the current stream API usage.

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                return null;
+            }
+            if (!nested.stream().allMatch(e -> e == null || e instanceof List)) {
+                throw new IllegalArgumentException("expected a nested array at level " + level);
+            }
+            List<?> head = (List<?>) nested.get(0);
+            if (head == null || head.isEmpty()) {
+                return null;
+            }
+            if (!nested.stream().allMatch(e -> e == null || ((List<?>) e).size() == head.size())) {
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:98]: The same comment about using the stream API. It is probably not that necessary to create a new stream object to iterate over a list and do some predicate check. You can achieve that with the for-loop in a not less concise manner as well.

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        this.boundSignature = boundSignature;
+    }
+
+    private static List<?> extractNestedArray(List<?> array, int targetLevel) {
+        List<?> nested = array;
+        for (int level = 2; level <= targetLevel; level++) {
+            if (nested == null || nested.isEmpty()) {
+                return null;
+            }
+            if (!nested.stream().allMatch(e -> e == null || e instanceof List)) {
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:91]: Same here, see the previous comments.

Comment (kovrus) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:91]: can we add tests that cover both scenarios?

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        return nested;
+    }
+
+    @Override
+    public Iterable<Row> evaluate(TransactionContext txnCtx, Input<T>... args) {
+        assert args.length == 2 || args.length == 3 : "number of args must be 2 or 3";
+
+        List<?> array = (List<?>) args[0].value();
+        if (array == null || array.isEmpty()) {
+            return () -> NO_RESULTS_ITERATOR;
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:112]: maybe align this variable. not sure why we need another static field for it.

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private GenerateSubscripts(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    private static List<?> extractNestedArray(List<?> array, int targetLevel) {
+        List<?> nested = array;
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:86]: minor: it is a bit confusing, `array` becomes `nested`?

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+                throw new IllegalArgumentException("nested arrays should have the same dimension within a level");
+            }
+            nested = head;
+        }
+        return nested;
+    }
+
+    @Override
+    public Iterable<Row> evaluate(TransactionContext txnCtx, Input<T>... args) {
+        assert args.length == 2 || args.length == 3 : "number of args must be 2 or 3";
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:108]: not really related, but i am wondering whether assertion like that still makes sense. isn't it already guarded by the func resolving/registration logic?

Comment (mfussenegger) [reply server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:108]: It is guarded by the function signature, but using assertions like this can be a useful defensive programming technique. 
It makes explicit that there exists some contract. It can help debugging in case the other component no longer ensures this.

But the assertion message could be improved. 

```suggestion
        assert args.length == 2 || args.length == 3 : "Signature must ensure that there are either two or three arguments"; 
```

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        for (int level = 2; level <= targetLevel; level++) {
+            if (targetArray == null || targetArray.isEmpty()) {
+                return null;
+            }
+            targetArray = checkAllElementsAreArraysOfConsistentDimension(targetArray, level);
+        }
+        return targetArray;
+    }
+
+    private static List<?> checkAllElementsAreArraysOfConsistentDimension(List<?> array, int level) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:93]: I think the name doesn't really fit. It implies that it verifies some condition, but also has a return value without indicating what the value is about.

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    @Override
+    public Iterable<Row> evaluate(TransactionContext txnCtx, Input<T>... args) {
+        assert args.length == 2 || args.length == 3 :
+            "Signature must ensure that there are either two or three arguments";
+
+        List<?> array = (List<?>) args[0].value();
+        if (array == null || array.isEmpty()) {
+            return () -> Collections.emptyIterator();
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:133]:
```suggestion
            return List.of();
```

No need to create a special supplier/`Iterable`.

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        }
+        int dim = (Integer) args[1].value();
+        if (dim <= 0) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:136]: `dim == null` case needs to be handled as well. This causes a `NullPointerException` as it is.

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+            throw new IllegalArgumentException("dimension must be greater than zero");
+        }
+        List<?> nestedArray = extractNestedArray(array, dim);
+        if (nestedArray == null || nestedArray.isEmpty()) {
+            return () -> Collections.emptyIterator();
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:141]:
```suggestion
            return List.of();
```

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+        }
+        boolean reversed = args.length == 3 && (Boolean) args[2].value();
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:143]: This raises a `NullPointerException` if the reversed flag argument is `null`

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+            public Row next() {
+                if (!hasNext()) {
+                    throw new NoSuchElementException("no more rows");
+                }
+                columns[0] = idx;
+                idx += step;
+                return row;
+            }
+        };
+    }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:167]: What do you think about extracting parts of the `evaluate` of `generate_series` and re-using it here?

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+import java.util.NoSuchElementException;
+
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+import static io.crate.types.DataTypes.INTEGER;
+import static io.crate.types.TypeSignature.parseTypeSignature;
+
+public final class GenerateSubscripts<T> extends TableFunctionImplementation<T> {
+
+    public static final FunctionName NAME = new FunctionName(PgCatalogSchemaInfo.NAME, "generate_subscripts");
+    private static final RowType RETURN_TYPE = new RowType(List.of(INTEGER), List.of("subscripts"));
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:49]: Is `subscripts` the right field-name?

File: docs/general/builtins/table-functions.rst [hunk 16]
```diff
@@ -140,6 +140,94 @@ The return value always matches the ``start`` / ``stop`` types.
...
+    | 4 |
+    | 3 |
+    | 2 |
+    | 1 |
+    +---+
+    SELECT 4 rows in set (... sec)
+
+The following examples work on an array of three dimensions. Each of the elements
+within a given level must be either NULL, or an array of the same size as the other
+arrays within the same level.
```
Comment (mfussenegger) [inline docs/general/builtins/table-functions.rst:188]: I wonder if we should actually be a bit more lenient about this and take the longest array at the given dimension?

In PostgreSQL this is a limitation of the array type itself:

```
=# select array[ array[1, 2],  array[3, 4, 5] ];
ERROR:  2202E: multidimensional arrays must have array expressions with matching dimensions
```

@kovrus what do you think?

File: docs/general/builtins/table-functions.rst [hunk 17]
```diff
@@ -140,6 +140,94 @@ The return value always matches the ``start`` / ``stop`` types.
...
+::
+
+    cr> select generate_subscripts([ [[1], null], [[3], [null]], [null, null]], 2) as s;
+    +---+
+    | s |
+    +---+
+    | 1 |
+    | 2 |
+    +---+
+    SELECT 2 rows in set (... sec)
```
Comment (mfussenegger) [inline docs/general/builtins/table-functions.rst:230]:
```suggestion
```


I think one example of the multi-dimensional case is enough. We don't need to cover all kind of edge cases or so for documentation purposes.

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+                }
+            }
+            targetArray = firstNonNullElement;
+        }
+        return targetArray != null ? targetArray.size() : 0;
+    }
+
+    @SafeVarargs
+    @Override
+    public final Bucket evaluate(TransactionContext txnCtx, Input<T>... args) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:124]: Why the change to `Bucket`? This can still be an Iterable, the additional size info of `Bucket` is not required or used.

File: server/src/main/java/io/crate/expression/tablefunctions/SingleColumnRangeIterator.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+import io.crate.data.Row;
+import io.crate.data.RowN;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+
+
+public final class SingleColumnRangeIterator<T, U extends Number> implements Iterator<Row> {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/tablefunctions/SingleColumnRangeIterator.java:35]: I think the `U extends Number` constraint is unecessary, no?

Also not sure about the name. I think it emphases the `SingleColumn` part too much. How about just `RangeIterator`, `SeriesGenerator`, or `GenerateSeriesIterator` ?

Also, could this an `Iterable`? Would make it a bit more powerful

File: server/src/main/java/io/crate/expression/tablefunctions/SingleColumnRangeIterator.java [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+    private final T endIncluded;
+    private final Function<T, T> stepMutator;
+    private final Comparator<T> comparator;
+    private final Function<T, U> resultExtractor;
+    private final boolean reversed;
+    private T idx;
+
+    public SingleColumnRangeIterator(T start,
+                                     T endIncluded,
+                                     Function<T, T> stepMutator,
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/tablefunctions/SingleColumnRangeIterator.java:48]:
```suggestion
                                     Function<T, T> step,
```

```suggestion
                                     Function<T, T> nextValue,
```

It's a pure function that shouldn't mutate any state.

File: server/src/main/java/io/crate/expression/tablefunctions/SingleColumnRangeIterator.java [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+    private final Object [] columns;
+    private final RowN row;
+    private final T endIncluded;
+    private final Function<T, T> stepMutator;
+    private final Comparator<T> comparator;
+    private final Function<T, U> resultExtractor;
+    private final boolean reversed;
+    private T idx;
+
+    public SingleColumnRangeIterator(T start,
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/tablefunctions/SingleColumnRangeIterator.java:46]:
```suggestion
    public SingleColumnRangeIterator(T startInclusive,
```

File: server/src/main/java/io/crate/expression/tablefunctions/SingleColumnRangeIterator.java [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+public final class SingleColumnRangeIterator<T, U extends Number> implements Iterator<Row> {
+
+    private final Object [] columns;
+    private final RowN row;
+    private final T endIncluded;
+    private final Function<T, T> stepMutator;
+    private final Comparator<T> comparator;
+    private final Function<T, U> resultExtractor;
+    private final boolean reversed;
+    private T idx;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/tablefunctions/SingleColumnRangeIterator.java:44]: Not sure if `idx` is the right name here. It is not an index into a range, but the actual value.

File: server/src/main/java/io/crate/expression/tablefunctions/SingleColumnRangeIterator.java (hunk 20 above)
Comment (marregui) [reply server/src/main/java/io/crate/expression/tablefunctions/SingleColumnRangeIterator.java:48]: Agreed, this is the fault of my background in clojure.

File: docs/general/builtins/table-functions.rst (hunk 16 above)
Comment (marregui) [reply docs/general/builtins/table-functions.rst:188]: The longest array would result  in `index out of bounds` errors.
The shortest array would result in missing elements.

File: server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private GenerateSubscripts(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    private static int getNumRows(List<?> array, int depthLevel) {
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java:78]: lgtm

it is a bit hard to follow sometime some var names and messages with them, we have dim, level, target level, depth level, etc. but i am not sure how it can be improved.
